x-pack/plugins/infra/public/containers/with_kuery_autocompletion.tsx

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { connect } from 'react-redux';
+
+import { AutocompleteSuggestion, getAutocompleteProvider } from 'ui/autocomplete_providers';
+import { StaticIndexPattern } from 'ui/index_patterns';
+
+import { sourceSelectors, State } from '../store';
+import { RendererFunction } from '../utils/typed_react';
+
+const withIndexPattern = connect((state: State) => ({
+  indexPattern: sourceSelectors.selectDerivedIndexPattern(state),
+}));
+
```

Comment:
Being new to this code base I had to stare at this file for a while until I understood how everything fits together. It all works perfectly, but I think we should consider staying closer to the plain, Dan Abramov tutorial-style way of putting together a container:
* actually have `mapStateToProps` and `mapDispatchToProps` as named functions that are passed to `connect()`, even if they are really straightforward and short
* move the call to `connect()` further down and use it in the conventional way where the function it returns is called immediately with the component to `connect` as a parameter

Alternatively, we could document the pattern that is used in this file (along with some others, like how action creators are generated) and stick to it everywhere, to reduce the cognitive overhead.

I know time is tight, so no immediate need to change anything for this PR, just my thoughts.

Reply:
Why I have an aversion to directly wrapping the component with `connect` is that it prohibits the independent use and testing of the component. In addition, since we are going for the function-as-children style of composing containers and presentational components, the `connect` alone does not cut it. I'm using the `asChildFunctionRenderer()` helper to automatically derive a well-typed function-as-child component from a `connect`ed HOC in most cases, but here we need additional component state beyond just redux (due to the interaction with the external autocompletion api).

I would be very interested to see alternative formulations of this container for comparison.

Reply:
What about the second suggestion then, documenting the patterns that are used and the reasons behind them?

Reply:
Certainly: #22452

And I really meant it when I wrote that I would be interested to read other formulations for these kinds of containers. This is certainly not the definitive way, just something type-safe that I came up with as a compromise between readability and DRYness.